### ISSUE TITLE
Diffs didn't match during apply: amazon elb

### ISSUE THREAD

**Issue by fishnix** [hashicorp/terraform#6773]
### Terraform Version

Terraform v0.6.15
### Affected Resource(s)
- aws_elb
### Terraform Configuration Files

``` hcl
resource "aws_instance" "portal_az1" {
  count = "${var.portal_az1_count}"
  instance_type = "${var.portal_az1_type}"
  ami = "${var.portal_az1_ami}"
  depends_on = [
    "aws_db_instance.primary",
    "aws_db_instance.replica",
    "aws_instance.redis_az1",
    "aws_instance.solr_az1"
  ]

  connection {
    user = "${var.connection_user}"
    key_file = "${var.key_path}"
    agent = false
  }

  key_name = "${var.key_name}"
  subnet_id = "${var.subnet_az1_id}"

  vpc_security_group_ids = ["${aws_security_group.environment.id}"]

  tags {
    Name = "Portal ${upper(var.environment)} AZ1 ${format("%02d", count.index+1)}"
    environment = "${var.environment}"
    application = "portal"
    container = "yajsw"
    deployment = "terraform"
    version = "${var.version_az1}"
  }
}

resource "aws_instance" "portal_az2" {
  count = "${var.portal_az2_count}"
  instance_type = "${var.portal_az2_type}"
  ami = "${var.portal_az2_ami}"
  depends_on = [
    "aws_db_instance.primary",
    "aws_db_instance.replica",
    "aws_instance.redis_az1",
    "aws_instance.solr_az1"
  ]

  connection {
    user = "${var.connection_user}"
    key_file = "${var.key_path}"
    agent = false
  }

  key_name = "${var.key_name}"
  subnet_id = "${var.subnet_az2_id}"

  vpc_security_group_ids = ["${aws_security_group.environment.id}"]

  tags {
    Name = "Portal ${upper(var.environment)} AZ2 ${format("%02d", count.index+1)}"
    environment = "${var.environment}"
    application = "portal"
    container = "yajsw"
    deployment = "terraform"
    version = "${var..version_az2}"
  }
}

resource "aws_elb" "portal" {
  name = "portal-${var.environment}-elb"
  internal = "false"
  connection_draining = "true"
  connection_draining_timeout = "300"
  idle_timeout = "300"

  listener {
    instance_port = 7878
    instance_protocol = "http"
    lb_port = 80
    lb_protocol = "http"
  }

  listener {
    instance_port = 7878
    instance_protocol = "http"
    lb_port = 443
    lb_protocol = "https"
    ssl_certificate_id = "${var.ssl_certificate_id}"
  }

  listener {
    instance_port = 7777
    instance_protocol = "http"
    lb_port = 7777
    lb_protocol = "http"
  }

  health_check {
    healthy_threshold = 2
    unhealthy_threshold = 3
    timeout = 5
    target = "HTTP:7777/ping"
    interval = 30
  }

  security_groups = [ "${aws_security_group.environment.id}",
                         "${aws_security_group.web.id}" ]

  subnets = ["${var.subnet_az1_id}", "${var.subnet_az2_id}"]

  instances = [ "${aws_instance.portal_az1.*.id}",
                "${aws_instance.portal_az2.*.id}"]

  tags {
    Name = "Portal ${upper(var.environment)} Load Balancer"
    environment = "${var.environment}"
    application = "portal"
  }
}

```
### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. `terraform plan`
2. `terraform apply`

Repeating the plan/apply succeeds

```
Error applying plan:

1 error(s) occurred:

* aws_elb.portal: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.6.15
    Resource ID: aws_elb.portal
    Mismatch reason: attribute mismatch: instances.1744977461
    Diff One (usually from plan): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"instances.1744977461":*terraform.ResourceAttrDiff{Old:"", New:"${aws_instance.portal_az2.*.id}", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "instances.2742378615":*terraform.ResourceAttrDiff{Old:"i-154e4488", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "instances.131389316":*terraform.ResourceAttrDiff{Old:"i-2468c7be", New:"i-2468c7be", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"instances.2742378615":*terraform.ResourceAttrDiff{Old:"i-154e4488", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "instances.131389316":*terraform.ResourceAttrDiff{Old:"i-2468c7be", New:"i-2468c7be", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "instances.374281886":*terraform.ResourceAttrDiff{Old:"", New:"i-0e627593", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.


Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by catsby** [comment]
Hey @fishnix an interesting find here, thanks for writing in!

Looking at the snippet you shared I immediately suspected the instances, as their new/old ids change. Was there any shifting of Instances in the background here? The `Old` and `New` don't match up across the plan/apply


Plan:
```
Old:"i-154e4488",
 New:"",

[...]

Old:"i-2468c7be",
 New:"i-2468c7be",
```

Apply:
```
Old:"i-2468c7be",
 New:"i-2468c7be",
[...]
Old:"",
 New:"i-0e627593",
```

Was this plan ran from an empty state, and you got the error on first plan->apply?

I was able to reproduce with this gist:

- https://gist.github.com/catsby/cac2fede751bcdace53a4fcbb99c9677

After some playing around and initially getting a similar diffs error, with a follow up plan giving a different error about availability zones for my subnets. What happened _for me_ was that both of my subnets were in the same AZ. Your config suggests your subnets are in different AZs, but I can't see that for sure because they are not in the configuration. Can you double check that for me? 

Lastly, I see in your config you have `var.environment`. You also have a security group reference like so: `aws_security_group.environment.id` . In my gist I included a security group named `environment`, but I'm curious, is that what you have? Or are you by chance trying to reference a different one based on an interpolation that isn't happening?

**Comment by fishnix** [comment]
Hey @catsby !  Sorry for the delay!

There are no instances changing behind the scenes here (I'm assuming you mean outside of terraform) and this happens every time we apply this code.  This isn't an empty state, although the code was mostly the same when it was started a couple months ago.  The state files are being stored remotely in S3 if that makes any difference.  These instances are definitely being created in different subnets in different AZs.

For the last question, no it's not failing to interpolate or anything, this is a module that was written to be reused and `environment` was just a poorly chosen identifier for the SG.  The environment var is  used to apply Name tags to all of those things, pick chef environments, etc.  

Hope that helps!  Thanks again!

**Comment by fishnix** [comment]
Good morning @catsby - just wanted to make sure you saw the reply.  Cheers!

**Comment by fishnix** [comment]
Hi! Any ideas on a work around for this issue?  It's stopping us from running terraform in automation scripts currently.  Thanks!

**Comment by jbardin** [comment]
Hi @fishnix:

I'm unfortunately unable to repro this at all, but I was wondering you could try your test case using the latest master version of terraform? I think the issue lies in the failed (or lack of) interpolation of an instance id, which should have been fixed.

**Comment by stack72** [comment]
Ping @fishnix - have you been able to try out the RC3 to see if your issue is fixed? :)

**Comment by cbarbour** [comment]
I've experienced the same issue and filed a bug a couple weeks back.

https://github.com/hashicorp/terraform/issues/8062

This has been reproduced on 0.7.0.

**Comment by fishnix** [comment]
I'm sorry @stack72 I just saw this, I'll check it out!

**Comment by joeig** [comment]
Same problem here after migrating from v0.6 to v0.7. The new variable type "list" might cause this issue.

**Comment by mitchellh** [comment]
This issue is fixed in 0.7.8
